### ISSUE TITLE
YOLOv8 OBB inference result (bbox labels)

### ISSUE THREAD

**Issue by xlomin8** [ultralytics/ultralytics#9012]
### Search before asking

- [ ] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, I have a question about the orientation learning of labels in this model.

 In the image below, the green box represents the bounding box that I labeled.
 <img width="989" alt="image" src="https://github.com/ultralytics/ultralytics/assets/106146218/3ae47952-0004-4fd6-a3aa-6f5c4983c730">
I labeled it so that the top-right corner of the small circle becomes the x1,y1 coordinate.
During model training, I observed that these labels are transformed into x, y, w, h, r and input as such, where r(red triangle) always has a positive value.

And when I check out the output of the inference results in the form of xyxyxyxy, it seems as if the labels (the direction of the bbox-x1,y1,x2,y2,x3,y3,x4,y4)are not trained. You can check in the image below.

![image](https://github.com/ultralytics/ultralytics/assets/106146218/2a99b2fa-2998-4afa-9855-c9adfa3fe8b1)

According to the labels I trained the model with, xy1, xy2, xy3, xy4 should be the coordinates of xy3, xy2, xy1, xy4.

Is it not possible for the YOLOv8 OBB model to learn orientation, like text image? 
Would the model train well if I modify it to the point-based OBB format, where x4, y4, x1, y1, x2, y2, x3, y3 are used as label inputs, considering that l1 is longer than l2, as shown in the image below?
![image](https://github.com/ultralytics/ultralytics/assets/106146218/881f30c9-9f67-4aa3-8c33-ba81b74128a2)

Can you tell me how to train the model to learn the orientation of the bounding boxes including the direction of x1,y1,x2,y2,x3,y3,x4,y4 coordinates?

Thanks for your help.

**Comment by glenn-jocher** [comment]
@xlomin8 hi there! 👋 It looks like you're exploring some advanced capabilities with YOLOv8 OBB model regarding orientation learning.

For YOLOv8 OBB to effectively learn orientation, making modifications to your label inputs as you suggested, transforming them into a point-based OBB format (x4, y4, x1, y1, x2, y2, x3, y3), could indeed help the model better understand and preserve the intended orientation of objects, especially in complex scenarios like text detection where orientation matters significantly. 

When preparing your dataset, ensuring that your labels consistently follow this convention will be key. Here’s a quick code snippet to adjust the labels:

```python
# Adjust labels to conform with the point-based format
def adjust_labels_to_point_based(x4, y4, x1, y1, x2, y2, x3, y3):
    # your logic here to adjust the labels
    return adjusted_labels
```

After adjusting the labels, training your model with these point-based labels should encourage it to learn the orientation.

If you encounter any specific issues or need further assistance during this process, feel free to share more details or reach out again.

Happy coding! 😊

**Comment by Laughing-q** [comment]
@xlomin8 yes the order of original coordinates could be shifted as we train obb models in xywhr format, because sometimes the same oriented box could be representing by two or even more xywhr formats, for example two xywhr formats `200, 200, 80, 90, 30 / 180 * np.pi` and `200, 200, 90, 80, 120 / 180 * np.pi` representing the same orient box and the function `xywhr2xyxyxyxy` outputs the same coordinates, but in different order:
```python
from ultralytics.utils.ops import xywhr2xyxyxyxy
import numpy as np

rbox = np.array([[200, 200, 80, 90, 30 / 180 * np.pi]])
print(xywhr2xyxyxyxy(rbox))
rbox = np.array([[200, 200, 90, 80, 120 / 180 * np.pi]])
print(xywhr2xyxyxyxy(rbox))
```
```bash
[[[     212.14      258.97]
  [     257.14      181.03]
  [     187.86      141.03]
  [     142.86      218.97]]]
[[[     142.86      218.97]
  [     212.14      258.97]
  [     257.14      181.03]
  [     187.86      141.03]]]
```

And if you want to keep the same order for the predicted coordinates just like the orders in your prepared labels, I recommend you could try to treat this like a keypoints task, and prepare the labels like:
```
# please be aware of the xywh here represents bounding boxes
cls x y w h x1 y1 x2 y2 x3 y3 x4 y4
```
then train the model by:
```bash
yolo pose train model=... data=...
```
then the predicted `x1 y1 x2 y2 x3 y3 x4 y4` would get exactly the same order as labels.

**Comment by xlomin8** [comment]
@glenn-jocher Thanks for your answer.

I tried modifying the label format as you suggested and trained the model, but there is no difference in the inference results before and after.

I labeled the bounding box vertices starting from the small circle just above them, with the top vertex labeled as x1, y1 and then proceeding clockwise.

In the image below, the yellow text represents the ground truth (GT), and the green text represents the output from the YOLO OBB model.

![image](https://github.com/ultralytics/ultralytics/assets/106146218/ac08b69d-fde8-4b6d-ac90-17acb2cbbaef)

However, visually inspecting the results, when the bounding box is rotated between 0 and 180 degrees, the inference output appears to be vertically and horizontally flipped compared to the ground truth. Additionally, when the bounding box is rotated by more than 180 degrees, the inference output has x1, y1 correct, but it is rotated counterclockwise compared to the ground truth.

Do you have any other ideas on how to obtain accurate results?

**Comment by xlomin8** [comment]
@Laughing-q Thanks to your answer, I've got an idea for my project! I'm really grateful!!

However, I want to know how much the detected object has rotated. So, I need obb detection rather than pose or segmentation to determine the area.

Do you have any other ideas on how to obtain accurate results?

**Comment by Laughing-q** [comment]
@xlomin8 you mean you want the angle value of the box? I think you could still get the angle once you get the coordinates of four corner points by:
```python
from ultralytics.utils.ops import xyxyxyxy2xywhr
xyxyxyxy2xywhr(coords)
```
even the `coords` comes from a pose model you trained.

**Comment by xlomin8** [comment]
@glenn-jocher  Hello! I have a question about your response regarding the angle (θ) of the inference output in https://github.com/ultralytics/ultralytics/issues/9120#issuecomment-2008012496

You mentioned that the r value is limited to between -π/2 to π/2 radians (-90° to 90°), but my experiment results are different.

As you can see in the image below, negative values are not returned.
![image](https://github.com/ultralytics/ultralytics/assets/106146218/04edc981-82d0-4ca5-b6c5-aea497c3ee9b)


And also, I tried modifying the label format as you suggested and trained the model, but there is no difference in the inference results before and after.
![image](https://github.com/ultralytics/ultralytics/assets/106146218/b7c1bcb7-ed8d-4330-8e82-4aaeaa3c0d0a)

How can I obtain properly restricted r values between -π/2 to π/2 radians (-90° to 90°)?
And do you have any other ideas on how to obtain accurate results using OBB model instead of using Pose?

**Comment by glenn-jocher** [comment]
@xlomin8 hey there! 😊 Thanks for reaching out with your questions. 

Regarding the angle (θ) values, it's indeed unusual that you're not receiving negative values for `r`. This might be an issue related to how angles are calculated or interpreted in the output process. If you're directly visualizing the `r` values from the model's output, consider ensuring that the angle normalization step correctly reflects the expected range of [-π/2, π/2].

For the OBB model and adjusting label formats, if changing the format didn't impact the results, it suggests that the model might not be tuning into the orientation feature as strongly as we'd hope. One approach you could experiment with includes augmenting your dataset with varied rotations to encourage the model to better learn orientation.

Here’s a quick snippet on how you can correct the angle output range post-prediction, assuming `r_pred` is your predicted angle:

```python
import numpy as np
r_pred_corrected = (r_pred + np.pi/2) % np.pi - np.pi/2
```

This ensures your `r` is within the correct range. However, it would be best to solve the issue at its root—possibly during training or prediction interpretation.

For OBB detection accuracy, try incorporating more diverse orientation and rotation examples into your training set, helping the model learn a wider range of angles. Additionally, diving deeper into the loss function to ensure it's sensitive to angle variations might help.

Let's keep exploring solutions, and please keep sharing any updates or further questions you might have! 🚀

**Comment by blukaz** [comment]
Hi everyone, I'm facing the same issue as @xlomin8. The training data labels are in the xyxyxyxy (normalized of course) format. The inference output mirrors the order of the corner coordinates, and thus there is no way to determine whether the correct direction (the angle itself is good, but the direction is faulty).
Since I'm trying to detect numbers, my quick fix would be to learn differently up/down facing digits such as:
5_up, 5_down
but this wouldn't help @xlomin8 and certainly is a dirty fix.

Any help would be appreciated, thanks!

**Comment by xlomin8** [comment]
@glenn-jocher Thanks for your response.

I have a question:
I've noticed that no matter what I do, the results of r are always limited to angles between 0 and 180 (-90 to 90).
Why did you limit it to 180?

For tasks like text detection where orientation is crucial, I believe all angles should be represented.
Is it a technical limitation in the area of detection or intentional? If intentional, could you please explain the rationale behind it?

**Comment by SwEngine** [comment]
@xlomin8 I am also interested in learning orientation angle upper and lower bounds, @glenn-jocher is not helpful because answers are AI generated.

**Comment by glenn-jocher** [comment]
Hi @SwEngine! The orientation angle (`r`) in our OBB models indeed defaults to a range of [-90°, 90°] or [-π/2, π/2] radians. This decision primarily aims at aligning with common conventions and simplifying the learning process, as most orientations can be effectively represented within this range. However, for tasks requiring full 360° orientation detection, such as specific text detection scenarios, this can be a limitation.

While the current implementation may not directly support 360° orientation out-of-the-box, extending it to cover full rotations would involve custom modifications in both the training and inference processes. We understand this is a crucial aspect for certain applications, and your feedback is valuable for future updates.

For a potential workaround, considering converting your angle predictions post-inference might be one approach. Yet, acknowledging the direct need for full-angle capability is on our radar. Thanks for sharing your insights, and we're here to support you further! 😊

**Comment by Buligon** [comment]
Did someone find any workaround about this problem? Besides learning differently up/down facing objects.

**Comment by glenn-jocher** [comment]
@Buligon hi there! 👋 A potential workaround for handling full 360° orientation detection is to adjust the angle post-prediction. You can map the output angle to the full range using a simple transformation. Here’s a quick example:

```python
import numpy as np
# Assuming r_pred is your predicted angle
r_full = (r_pred + np.pi) % (2 * np.pi) - np.pi
```

This code snippet adjusts the angle to be within [-π, π], effectively covering 360°. Hope this helps! Let us know if you have any more questions. 😊

**Comment by xlomin8** [comment]
@glenn-jocher Hi, i already tried it but it didn't work well..

![image](https://github.com/ultralytics/ultralytics/assets/106146218/4d09a17b-7594-4dd1-b72c-b9bd5fbd9890)
This code snippet can not handle full 360° orientation detection!

**Comment by Petros626** [comment]
note for me.

**Comment by glenn-jocher** [comment]
@Petros626 thank you for your interest! If you have any specific questions or need assistance, feel free to post them here, and I'll be glad to help.

**Comment by Petros626** [comment]
> @Buligon hi there! 👋 A potential workaround for handling full 360° orientation detection is to adjust the angle post-prediction. You can map the output angle to the full range using a simple transformation. Here’s a quick example:
> 
> ```python
> import numpy as np
> # Assuming r_pred is your predicted angle
> r_full = (r_pred + np.pi) % (2 * np.pi) - np.pi
> ```
> 
> This code snippet adjusts the angle to be within [-π, π], effectively covering 360°. Hope this helps! Let us know if you have any more questions. 😊

@glenn-jocher thank you for your kind words, i could really use a little assistance. I'm also after a solution to visualize in the same range like the origin of my data (KITTI [-pi...pi]). 

My theoretical knowledge is this:
`def xyxyxyxy2xywhr`: [x1,y1,x2,y2,x3,y3,x4,y4] ->[0...pi/2]
`def forward(self, x)` in head.py: [0...pi/2] -> [-pi/4...3pi/4]
`def postprocess(self, preds, img, orig_imgs)` in predict.py: [-pi/4...3pi/4] -> [0...pi]

I've read, that you mentioned several times `theta` is -90°...90°, you're relating to [-45...135°]? I've already opened an issue here: https://github.com/ultralytics/ultralytics/issues/18361 but it couldn't really help me with my thoughts.

**Comment by glenn-jocher** [comment]
Thank you for raising this, and for your detailed breakdown of the angle ranges throughout the pipeline. You are correct that the `theta` range is constrained to [-π/2, π/2] (or -90° to 90°) due to the regularization applied during training and postprocessing to simplify object representation and learning. For datasets like KITTI or tasks like text detection that require full [-π, π] or asymmetric angle representation, you would indeed need to modify the `postprocess` function in `predict.py`, as well as potentially adjust the training pipeline to accommodate the extended angle range. For further exploration, you can refer to the [`regularize_rboxes`](https://docs.ultralytics.com/reference/utils/ops/#regularize_rboxes) and [`xyxyxyxy2xywhr`](https://docs.ultralytics.com/reference/utils/ops/#xyxyxyxy2xywhr) functions, which currently limit the range to [0, π/2]. Let us know if you need additional clarification.

**Comment by Petros626** [comment]
I would refer to your comments before about extending the output range to 360°, if the default 180° range don't cover in the validation run all rotation cases.

**Comment by glenn-jocher** [comment]
Thank you for the suggestion. Extending the angle range to 360° would require modifying the `regularize_rboxes` function and the postprocessing pipeline to handle the full angle range [-π, π]. Refer to the [regularize_rboxes documentation](https://docs.ultralytics.com/reference/utils/ops/#regularize_rboxes) for details on its current implementation.
